src/Proto3/Suite/Class.hs

```diff
@@ -485,51 +484,61 @@ instance MessageField BL.ByteString
 instance (Named e, ProtoEnum e) => MessageField (Enumerated e)
 
 instance (Ord k, Primitive k, MessageField k, Primitive v, MessageField v) => MessageField (M.Map k v) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1)) . M.toList
+  encodeMessageField num =
+    let encodeField :: k -> v -> Encode.MessageBuilder
+        encodeField k v = Encode.embedded num (encodeMessage 1 (k, v))
+     in M.foldrWithKey (\k -> (<>) . encodeField k) mempty
 
   -- Data.Map.fromList will retain the last key/value mapping. From the spec:
   --
   -- > When parsing from the wire or when merging, if there are duplicate map
   -- > keys the last key seen is used.
-  decodeMessageField = M.fromList . Foldable.toList
-                       <$> repeated (Decode.embedded' (decodeMessage (fieldNumber 1)))
+  decodeMessageField = M.fromList <$> repeated (Decode.embedded' (decodeMessage 1))
+
   protoType _ = messageField (Map (primType (proxy# :: Proxy# k)) (primType (proxy# :: Proxy# v))) Nothing
 
 instance {-# OVERLAPS #-} (Ord k, Primitive k, Named v, Message v, MessageField k) => MessageField (M.Map k (Nested v)) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1)) . M.toList
+  encodeMessageField num =
+    let encodeField :: k -> Nested v -> Encode.MessageBuilder
+        encodeField k v = Encode.embedded num (encodeMessage 1 (k, v))
+     in M.foldrWithKey (\k -> (<>) . encodeField k) mempty
 
   -- Data.Map.fromList will retain the last key/value mapping. From the spec:
   --
   -- > When parsing from the wire or when merging, if there are duplicate map
   -- > keys the last key seen is used.
-  decodeMessageField = M.fromList . Foldable.toList
-                       <$> repeated (Decode.embedded' (decodeMessage (fieldNumber 1)))
+  decodeMessageField = M.fromList <$> repeated (Decode.embedded' (decodeMessage 1))
+
   protoType _ = messageField (Map (primType (proxy# :: Proxy# k)) (Named . Single $ nameOf (proxy# :: Proxy# v))) Nothing
 
 instance (HasDefault a, Primitive a) => MessageField (ForceEmit a) where
   encodeMessageField = encodePrimitive
 
 instance (Named a, Message a) => MessageField (Nested a) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1))
-                           . coerce @(Nested a) @(Maybe a)
-  decodeMessageField = coerce @(Parser RawField (Maybe a)) @(Parser RawField (Nested a))
-                       (Decode.embedded (decodeMessage (fieldNumber 1)))
+  encodeMessageField num x =
+    case coerce @(Nested a) @(Maybe a) x of
+      Nothing -> mempty
+      Just x' -> Encode.embedded num (encodeMessage 1 x')
+
+  decodeMessageField = coerce @(Maybe a) @(Nested a) <$> Decode.embedded (decodeMessage 1)
```

Comment:
`fmap` of a coercion can be slower than just a coercion.  Perhaps in this case it is the same speed because the `Functor` in question is:
```
newtype Parser input a = Parser { runParser :: input -> Either ParseError a }
```
Also, that newtype constructor is indeed imported into your module.  So the compiler might be able to see that there's no need to actually pattern-match on the `Either` just to coerce its `Right`.

But still, the old code made clear that the coercion was zero-cost, and for that reason I would prefer the old code in this particular case.

Reply:
Didn't know GHC inspects patterns when simplifying coercions, I thought it always erased them.

Reply:
The point is that without knowing that a `Parser` is, GHC must hold out the possibility that `fmap coerce` does more than just coercion.  That is, `fmap` might have a side effect unrelated to the argument that you pass.